### PR TITLE
Remove NodeBuilder

### PR DESCRIPTION
**Pull request by rjernst**
The NodeBuilder is currently used to construct a Node. However, this is
really just yet-another-builder that wraps around a Settings.Builder
witha couple convenience methods. But there are very few uses of these
convenience methods.  This change removes NodeBuilder, in favor of just
using the Node constructor.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/benchmark/get/SimpleGetActionBenchmark.java [hunk 1]
```diff
@@ -31,11 +31,11 @@
     public static void main(String[] args) {
         long OPERATIONS = SizeValue.parseSizeValue("300k").singles();
 
-        Node node = NodeBuilder.nodeBuilder().node();
+        Node node = new Node(Settings.EMPTY);
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/benchmark/get/SimpleGetActionBenchmark.java]: .start() missing

File: core/src/test/java/org/elasticsearch/benchmark/breaker/CircuitBreakerBenchmark.java [hunk 2]
```diff
@@ -123,7 +122,7 @@ public void run() {
     }
 
     public static void main(String args[]) throws Exception {
-        Node node = NodeBuilder.nodeBuilder().settings(Settings.settingsBuilder()).node();
+        Node node = new Node(Settings.EMPTY);
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/benchmark/breaker/CircuitBreakerBenchmark.java]: .start() missing

File: core/src/test/java/org/elasticsearch/benchmark/search/geo/GeoDistanceSearchBenchmark.java [hunk 3]
```diff
@@ -41,7 +40,8 @@
 
     public static void main(String[] args) throws Exception {
 
-        Node node = NodeBuilder.nodeBuilder().clusterName(GeoDistanceSearchBenchmark.class.getSimpleName()).node();
+        Settings settings = Settings.builder().put("cluste.name", GeoDistanceSearchBenchmark.class.getSimpleName()).build();
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/benchmark/search/geo/GeoDistanceSearchBenchmark.java]: typo: cluste.name -> cluster.name

**Comment by ywelsch** [comment]
Left some minor comments (missing `start()` calls and typo), o.w. LGTM.

**Comment by s1monw** [comment]
+1

**Comment by jasontedor** [comment]
I'm in favor of this change, but not in favor of making a change that will break compilation of client code in a minor release (this is currently labeled v2.2.0).

**Comment by rjernst** [comment]
@jasontedor NodeBuilder has always been an internal API. The fact that some users (who embed ES) use it instead of the Node constructor is not a reason we should hold back internal refactoring. Users who embed ES are advanced enough that they can update to use the Node constructor.

**Comment by rjernst** [comment]
I removed the 2.2 label, and opened an issue (#15383) to change the docs so constructing an in memory Node is never recommended.

**Comment by s1monw** [comment]
> I'm in favor of this change, but not in favor of making a change that will break compilation of client code in a minor release (this is currently labeled v2.2.0).

Since we are in java you have to recompile your client code anyway so the compiler will tell you what's going on. This is a great way to move on with APIs - when I am in the situation I was always happy to see a hard break to make the changes early and often.

**Comment by clintongormley** [comment]
@rjernst can we have a note in the breaking changes docs please

**Comment by s1monw** [comment]
> @rjernst can we have a note in the breaking changes docs please
is this necessary? It's unsupported really and a Java API

**Comment by dadoonet** [comment]
Well. It has always been documented in the Java API doc. So that makes it a bit "official" IMO.

**Comment by bleskes** [comment]
It's how we now officially communicate on how to start a client.  https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/node-client.html 

+1 on clear messaging as to why we are changing it - to me, breaking changes feel natural.

**Comment by mikemccand** [comment]
+1 to freely break this in 2.x and explain why in the breaking changes docs: javac makes it completely clear to these very expert users on upgrade that they need to use the `Node` constructor instead.

**Comment by bleskes** [comment]
> these very expert users 

I think this is the source of where people have different feelings. These are not expert settings - it's what we recommend on our website and many Java people use. I'm totally +1 with changing that (on 3.0 I would go as far as not treating this as a client at all). We should clear document this, explain why and offer people an alternative.

**Comment by 1ambda** [comment]
I think this is missing in 5.0 API breaking changes. 

https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking-changes-5.0.html

**Comment by clintongormley** [comment]
> I think this is missing in 5.0 API breaking changes.

True.  @rjernst could you add to the breaking changes please?

**Comment by rjernst** [comment]
Added:
5.x: 71bc4c6
5.0: bc5d7a9

**Pull request merged**